File: extract-frames/extract-frames.cpp
#include "stdafx.h"
#include <opencv2/opencv.hpp>
#include <ostream>
#include <iostream>

using namespace std;
using namespace cv;

int main(int argc, char** argv)
{
  if (argc != 3)
  {
    cout << "Usage: extract-frames.exe <video> <frames>" << endl;
    return -1;
  }

  auto inputVideoFilename = argv[1];
  auto framesToExtract = atoi(argv[2]);

  VideoCapture cap(inputVideoFilename);

  if (!cap.isOpened())
  {
    cout << "Cannot open the video file" << endl;
    return -1;
  }

  auto frameCount = cap.get(CV_CAP_PROP_FRAME_COUNT);
  auto frameCounter = 0;

  for (auto i = 0; i < frameCount; i++)
  {
    Mat frame;
    auto success = cap.read(frame);
    if (!success)
    {
      cout << "Cannot read frame " << to_string(i) << endl;
      continue;
    }

    if (i == 0 || i % static_cast<int>(frameCount / framesToExtract) != 0)
    {
      // do not extract every frame, but once in a while so we have a fixed number of frames
      // not correlated to the frame count
      continue;
    }

    auto filename = "Frame " + to_string(++frameCounter) + ".jpg";

    imwrite(filename, frame);
    cout << "saved " << filename << endl;
  }
}

File: src/main/java/ch/sebastianhaeni/thermotrains/internals/ExtractFrames.java
package ch.sebastianhaeni.thermotrains.internals;

import java.util.function.Function;
import java.util.function.Predicate;

import javax.annotation.Nonnull;

import ch.sebastianhaeni.thermotrains.util.Direction;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.opencv.core.Mat;
import org.opencv.videoio.VideoCapture;
import org.opencv.videoio.Videoio;

import static ch.sebastianhaeni.thermotrains.util.FileUtil.emptyFolder;
import static ch.sebastianhaeni.thermotrains.util.FileUtil.saveMat;
import static org.opencv.core.Core.flip;

public final class ExtractFrames {

  private static final Logger LOG = LogManager.getLogger(ExtractFrames.class);

  private ExtractFrames() {
    // nop
  }

  public static void extractFrames(
    @Nonnull String inputVideoFilename,
    @Nonnull String outputFolder) {

    extractFrames(inputVideoFilename, outputFolder, Direction.FORWARD, 50);
  }

  static void extractFrames(
    @Nonnull String inputVideoFilename,
    @Nonnull String outputFolder,
    @Nonnull Direction direction,
    int framesToExtract) {

    extractFrames(inputVideoFilename, outputFolder, direction, framesToExtract, 1);
  }

  /**
   * Extract n frames in a direction from an input file. The lengthFactor gets multiplied with the video length and only
   * the frames from start this amount of frames will be considered.
   */
  static void extractFrames(
    @Nonnull String inputVideoFilename,
    @Nonnull String outputFolder,
    @Nonnull Direction direction,
    int framesToExtract,
    double lengthFactor) {

    emptyFolder(outputFolder);

    VideoCapture capture = new VideoCapture();

    if (!capture.open(inputVideoFilename)) {
      throw new IllegalStateException("Cannot open the video file");
    }

    boolean isForward = direction == Direction.FORWARD;
    int frameCount = (int) (capture.get(Videoio.CAP_PROP_FRAME_COUNT) * lengthFactor);
    int framesBetween = frameCount / framesToExtract;
    int frameCounter = 0;

    Predicate<Integer> termination = isForward ?
      i -> i < frameCount :
      i -> i > 0;
    Function<Integer, Integer> increment = isForward ?
      i -> i + 1 :
      i -> i - 1;

    int i = isForward ? 0 : frameCount;

    while (termination.test(i) && frameCounter < framesToExtract) {
      i = increment.apply(i);

      Mat frame = new Mat();
      boolean success = capture.read(frame);
      if (!success) {
        LOG.warn("Cannot read frame {}", i);
        continue;
      }

      if (i == 0 || i % framesBetween != 0) {
        // do not extract every frame, but once in a while so we have a fixed number of frames
        // not correlated to the frame count
        continue;
      }

      if (!isForward) {
        flip(frame, frame, 1);
      }

      saveMat(outputFolder, frame, ++frameCounter);
    }
  }
}
